config/deprecation.go

```diff
@@ -0,0 +1,362 @@
+package config
+
+import (
+	"fmt"
+	"log" //nolint:revive // log is ok here as the logging facility is not set-up yet
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/fatih/color"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/aggregators"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/outputs"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+// Escalation level for the plugin or option
+type Escalation int
+
+func (e Escalation) String() string {
+	switch e {
+	case Warn:
+		return "WARN"
+	case Error:
+		return "ERROR"
+	}
+	return "NONE"
+}
+
+const (
+	// None means no deprecation
+	None Escalation = iota
+	// Warn means deprecated but still within the grace period
+	Warn
+	// Error means deprecated and beyond grace period
+	Error
+)
+
+// deprecationInfo contains all important information to describe a deprecated entity
+type deprecationInfo struct {
+	// Name of the plugin or plugin option
+	Name string
+	// Level of deprecation
+	Level Escalation
+	info  telegraf.DeprecationInfo
+}
+
+func (di *deprecationInfo) determineEscalation(major, minor int) {
+	var removalMajor, removalMinor int
+
+	di.Level = None
+	if di.info.Since == "" {
+		return
+	}
+
+	sinceMajor, sinceMinor := ParseVersion(di.info.Since)
+	if di.info.RemovalIn != "" {
+		removalMajor, removalMinor = ParseVersion(di.info.RemovalIn)
+	} else {
+		removalMajor, removalMinor = sinceMajor+1, 0
+		di.info.RemovalIn = fmt.Sprintf("%d.%d", removalMajor, removalMinor)
+	}
+
+	di.Level = None
+	if major > removalMajor || (major == removalMajor && minor >= removalMinor) {
+		di.Level = Error
+	} else if major > sinceMajor || (major == sinceMajor && minor >= sinceMinor) {
+		di.Level = Warn
+	}
```

Comment:
What about [go-semver](https://github.com/coreos/go-semver)?

Reply:
Didn't know about this lib. Will switch to it.